specs/introduction.md

```diff
@@ -13,3 +13,63 @@ To simplify we separate our proofs into two components.
 2. **EVM proof**: This checks that the correct opcode is called at the correct time. It checks the validity of these opcodes. It also confirms that for each of these opcodes the state proof performed the correct operation.
 
 Only after verifying both proofs are we confident that that Ethereum block is executed correctly.
+
+
+## Bus mapping
+
+
+Our state proof and EVM proof need to talk to eachother. We need some kind of accumulator to do this. Merkle trees where for a long time the best way of achiveing such communication. But this would add 1000’s of consraints to each opcode. instead we use a plookup key value mapping. We call this bus mapping afte the bus on a traditional computer.
+
+One of the big challenges of building a snark to verify a VM is that you need to be able to read random values / opcodes at any time. To do this we use key-value mappings. Key value mappings are basically random linear combinations of all elements
+
+
+### Plookup key-value mappings
+In plookup you can build key-value mappings as follows
+
+``` python
+def build_mapping():
+    keys = [1,3,5]
+    values = [2,4,6]
+
+    randomness = hash(keys, values)
+
+    mappings = []
+
+    for key , value in zip(keys,values):
+        mappings.append(key + randomness*value)
+    return(mappings)
+```
+
+
+It can be opened at the cost of 3 plookups
+
+``` python
+def open_mapping(mappings, keys, values):
+    randomness = hash(keys,values)
+    index = 1
```

Comment:
When we open them in the circuit, do we open 1 key and value at a time? Or do we open all keys and values?

How do we know the index?

Reply:
One key and value at a time. 

If we want to index we must add another value 

index + randomness* key + randomness**2 * value.